integration/simulation/simulation_test.go

```diff
@@ -7,68 +7,172 @@ import (
 	rand.Seed(time.Now().UnixNano())
 	uuid.EnableRandPool()
 
-	// create a folder specific for the test
-	err := os.MkdirAll("../.build/simulations/", 0o700)
-	if err != nil {
-		panic(err)
+	logFile := setupTestLog("../.build/simulations/")
+	defer logFile.Close()
+
+	// define core test parameters
+	numberOfNodes := 10
+	simulationTimeSecs := 15                // in seconds
+	avgBlockDurationUSecs := uint64(20_000) // in u seconds 1 sec = 1e6 usecs
+	avgLatency := avgBlockDurationUSecs / 15
+	avgGossipPeriod := avgBlockDurationUSecs / 3
+
+	// define network params
+	stats := NewStats(numberOfNodes)
+	l1NetworkConfig := NewL1Network(avgBlockDurationUSecs, avgLatency, stats)
+	l2NetworkCfg := NewL2Network(avgBlockDurationUSecs, avgLatency)
+
+	// define instances of the simulation mechanisms
+	txManager := NewTransactionManager(5, l1NetworkConfig, l2NetworkCfg, avgBlockDurationUSecs, stats)
+	simulation := NewSimulation(
+		numberOfNodes,
+		l1NetworkConfig,
+		l2NetworkCfg,
+		avgBlockDurationUSecs,
+		avgGossipPeriod,
+		stats,
+	)
+
+	// execute the simulation
+	simulation.Start(txManager, simulationTimeSecs)
+
+	// run tests
+	checkBlockchainValidity(t, txManager, simulation)
+
+	// generate and print the final stats
+	t.Logf("%+v\n", NewOutputStats(simulation))
+	// pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+}
+
+func checkBlockchainValidity(t *testing.T, txManager *TransactionManager, network *Simulation) {
+	// TODO check all nodes are the same height ?
+	// pick one node to draw height
+	l1Node := network.l1Network.nodes[0]
+	obscuroNode := network.l2Network.nodes[0]
+	currentBlockHead := obscuroNode.DB().GetCurrentBlockHead()
+	currentRollupHead := obscuroNode.DB().GetCurrentRollupHead()
+
+	l1Height := currentBlockHead.Height
+	l1HeightHash := currentBlockHead.ID
+	l2Height := currentRollupHead.Height
+
+	// ensure the L1 blocks are valid
+	validateL1(t, network.l1Network.Stats, l1Height, &l1HeightHash, l1Node)
+
+	// ensure the validity of l1 vs l2 stats
+	validateL1L2Stats(t, obscuroNode, network.l1Network.Stats)
+
+	// ensure the generated withdrawal stats match the l2 blockchain state (withdrawals)
+	totalWithdrawn := validateL2WithdrawalStats(t, obscuroNode, network.l1Network.Stats, l2Height, txManager)
+
+	// ensure that each node has the expected total balance computed above
+	validateL2NodeBalances(t, network.l2Network, network.l1Network.Stats, totalWithdrawn, txManager.wallets)
+
+	// ensure that each node can fetch each of the generated transactions
+	validateL2TxsExist(t, network.l2Network.nodes, txManager)
+}
+
+// validateL1L2Stats validates blockchain wide properties between L1 and the L2
+func validateL1L2Stats(t *testing.T, node *obscuro_node.Node, stats *Stats) {
+	l1Height := uint(0)
+	for header := node.DB().GetCurrentBlockHead(); header != nil; header = node.DB().GetBlockHeader(header.Parent) {
+		l1Height++
 	}
-	fileName := fmt.Sprintf("simulation-result-%d-*.txt", time.Now().Unix())
```

Comment:
this got lost in the merge.

Reply:
It's in https://github.com/obscuronet/obscuro-playground/blob/c4cb4f5424aa9abd729dcbdf7b4ba25f28d95e41/integration/simulation/simulation_test.go#L293

:)

Reply:
Joel changed it yesterday so that the random part is the nr of millis since start of epoch.
It's better because now the log files are ordered.

Reply:
Gotcha, I thought you meant it was deleted. 👍